File: 0752_openTheLock/main.cpp
#include <vector>
#include <string>
#include <unordered_set>
#include <queue>

class Solution {
public:
    int openLock(std::vector<std::string>& deadends, std::string target) {
        std::unordered_set<std::string> uset(deadends.begin(), deadends.end());
        std::queue<std::string> myQueue;
        if(uset.find("0000") != uset.end()) return -1;
        myQueue.push("0000");
        uset.insert("0000");
        int step = 0;
        while(!myQueue.empty()) {
            for(int i = myQueue.size(); i > 0; --i) {
                std::string cur = myQueue.front();
                myQueue.pop();
                if(cur == target) return step;
                for(int str_idx = 0; str_idx < 4; ++str_idx) {
                    std::string down = cur;
                    down[str_idx] = (down[str_idx] - '0' - 1 + 10) % 10 + '0';
                    if(uset.find(down) == uset.end()){
                        myQueue.push(down);
                        uset.insert(down);
                    }
                    std::string up = cur;
                    up[str_idx] = (up[str_idx] - '0' + 1 + 10) % 10 + '0';
                    if(uset.find(up) == uset.end()) {
                        myQueue.push(up);
                        uset.insert(up);
                    }
                }
            }
            ++step;
        }
        return -1;
    }
};

int main() {
    std::vector<std::string> test = {"0201","0101","0102","1212","2002"};
    std::string target = "0202";
    
    Solution sol;
    sol.openLock(test, target);

    return 0;
}

File: 0752_openTheLock/main.py
from typing import List
import collections

class Solution:
    def openLock(self, deadends: List[str], target: str) -> int:
        deadends_set = set(deadends)
        if "0000" in deadends_set:
            return -1
        
        myQueue = collections.deque(["0000"])
        step = 0
        while myQueue:
            for i in range(len(myQueue)):
                cur = myQueue.popleft()
                if cur == target:
                    return step
                for str_idx in range(4):
                    for inc in (-1, 1):
                        temp = str((int(cur[str_idx]) + inc + 10) % 10)
                        new_cur = cur[:str_idx] + temp + cur[str_idx+1:]
                        
                        if new_cur not in deadends_set:
                            deadends_set.add(new_cur)
                            myQueue.append(new_cur)
            step += 1
        return -1

test = ["0201","0101","0102","1212","2002"]
target = "0202"
sol = Solution()
sol.openLock(test, target)